src/swdmt/redistricting/StarGeneratorTest.java

```diff
@@ -0,0 +1,36 @@
+package swdmt.redistricting;
+import org.junit.jupiter.api.Test;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+class StarGeneratorTest {
+    /**
+     * Ensure each pattern generated has the correct number of Locations,
+     * and there are no duplicates in any of the generated patterns.
+     */
+    @Test
+    void generatePatternTest() {
+        int n = 100;
+        ArrayList<ArrayList<Location>> patterns = StarGenerator.generatePattern(n);
+        for(ArrayList<Location> pattern: patterns) {
+            assertEquals(pattern.size(), n);
+            Set<Location> patternSet = new HashSet<>();
+            Location previousDistrict = null;
+            for(Location currentDistrict: pattern) {
+                if(previousDistrict != null) {
+                    assertTrue(
+                            currentDistrict.xCoordinate() + 1 == previousDistrict.xCoordinate() ||
+                                    currentDistrict.xCoordinate() - 1 == previousDistrict.xCoordinate() ||
+                                    currentDistrict.yCoordinate() + 1 == previousDistrict.yCoordinate() ||
+                                    currentDistrict.yCoordinate() - 1 == previousDistrict.yCoordinate()
+                    );
+                }
+                previousDistrict = currentDistrict;
+            }
+            pattern.forEach(location -> patternSet.add(location));
+            assertEquals(patternSet.size(), pattern.size());
```

Comment:
What is the meaning of this code? How does it test StarGenerator?

It appears to confirm that the [`size()`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ArrayList.html#size()) method of [ArrayList](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ArrayList.html) returns the value specified in the API: "Returns the number of elements in this list."

Perhaps this code needs additional explanation to get at some subtlety that isn't intuitively obvious.